core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -796,16 +806,57 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       .asScala
       .toList
     stale.foreach { log =>
-      if (log.appId.isEmpty) {
+      if (log.appId.isEmpty &&
+          (!conf.get(DRIVER_LOG_DFS_DIR).isDefined ||
```

Comment:
This check is kinda awkward. How about a new property in `LogInfo` with the type of the log?

You could then also just filter those out above before `.toList`, since the `clearBlacklist` call is unrelated to the logs you're adding.